common/watcher/watcher.go

```diff
@@ -0,0 +1,182 @@
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/openmfp/golang-commons/logger"
+)
+
+// FileEventHandler handles file system events
+type FileEventHandler interface {
+	OnFileChanged(filepath string)
+	OnFileDeleted(filepath string)
+}
+
+// FileWatcher provides common file watching functionality
+type FileWatcher struct {
+	watcher *fsnotify.Watcher
+	handler FileEventHandler
+	log     *logger.Logger
+}
+
+// NewFileWatcher creates a new file watcher
+func NewFileWatcher(handler FileEventHandler, log *logger.Logger) (*FileWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create file watcher: %w", err)
+	}
+
+	return &FileWatcher{
+		watcher: watcher,
+		handler: handler,
+		log:     log,
+	}, nil
+}
+
+// WatchSingleFile watches a single file with debouncing
+func (w *FileWatcher) WatchSingleFile(ctx context.Context, filePath string, debounceMs int) error {
+	if filePath == "" {
+		w.log.Info().Msg("no file path provided, skipping file watcher")
+		<-ctx.Done()
+		return ctx.Err()
+	}
+
+	// Watch the directory containing the file
+	fileDir := filepath.Dir(filePath)
+	if err := w.watcher.Add(fileDir); err != nil {
+		return fmt.Errorf("failed to watch directory %s: %w", fileDir, err)
+	}
+	defer w.watcher.Close()
+
+	w.log.Info().Str("filePath", filePath).Msg("started watching file")
+
+	return w.watchWithDebounce(ctx, filePath, time.Duration(debounceMs)*time.Millisecond)
+}
+
+// WatchDirectory watches a directory recursively without debouncing
+func (w *FileWatcher) WatchDirectory(ctx context.Context, dirPath string) error {
+	// Add directory and subdirectories recursively
+	if err := w.addWatchRecursively(dirPath); err != nil {
+		return fmt.Errorf("failed to add watch paths: %w", err)
+	}
+	defer w.watcher.Close()
+
+	w.log.Info().Str("dirPath", dirPath).Msg("started watching directory")
+
+	return w.watchImmediate(ctx)
+}
+
+// watchWithDebounce handles events with debouncing for single file watching
+func (w *FileWatcher) watchWithDebounce(ctx context.Context, targetFile string, debounceDelay time.Duration) error {
+	var debounceTimer *time.Timer
+
+	for {
+		select {
+		case <-ctx.Done():
+			w.log.Info().Msg("stopping file watcher")
+			if debounceTimer != nil {
+				debounceTimer.Stop()
+			}
+			return ctx.Err()
+
+		case event, ok := <-w.watcher.Events:
+			if !ok {
+				return fmt.Errorf("file watcher events channel closed")
+			}
+
+			if w.isTargetFileEvent(event, targetFile) {
+				w.log.Debug().Str("event", event.String()).Msg("file changed")
+
+				// Simple debouncing: cancel previous timer and start new one
+				if debounceTimer != nil {
+					debounceTimer.Stop()
+				}
+				debounceTimer = time.AfterFunc(debounceDelay, func() {
+					w.handler.OnFileChanged(targetFile)
+				})
+			}
+
+		case err, ok := <-w.watcher.Errors:
+			if !ok {
+				return fmt.Errorf("file watcher errors channel closed")
```

Comment:
There could still be a `debounceTimer` active. maybe consider moving the 
```
			if debounceTimer != nil {
				debounceTimer.Stop()
			}
```
into a top level defered function.